airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io/airbyte/cdk/read/cdc/Debezium.kt

```diff
@@ -0,0 +1,111 @@
+        if (!wrapped.has(fieldName)) {
+            return NullNode.getInstance()
+        }
+        return wrapped[fieldName]
+    }
+}
+
+/** [DebeziumRecordValue] wraps a Debezium change data event value. */
+@JvmInline
+value class DebeziumRecordValue(val wrapped: JsonNode) {
+
+    /** True if this is a Debezium heartbeat event. These aren't passed to [DebeziumConsumer]. */
+    val isHeartbeat: Boolean
+        get() = source.isNull
+
+    /** The datum prior to this event; null for insertions. */
+    val before: JsonNode
+        get() = element("before")
+
+    /** The datum following this event; null for deletions. */
+    val after: JsonNode
+        get() = element("after")
+
+    /** Metadata containing transaction IDs, LSNs, etc; null for heartbeats. */
+    val source: JsonNode
+        get() = element("source")
+
+    val operation: String?
+        get() = element("op").takeUnless { it.isNull }?.asText()
+
+    /** Convenience function for accessing child object nodes of the debezium event root node. */
+    fun element(fieldName: String): JsonNode {
+        if (!wrapped.has(fieldName)) {
+            return NullNode.getInstance()
+        }
+        return wrapped[fieldName]
+    }
+}
+
+/** [DebeziumOffset] wraps the contents of a Debezium offset file. */
+@JvmInline value class DebeziumOffset(val wrapped: Map<JsonNode, JsonNode>)
+
+/** [DebeziumSchemaHistory] wraps the contents of a Debezium schema history file. */
+@JvmInline value class DebeziumSchemaHistory(val wrapped: List<HistoryRecord>)
+
+/** Debezium Engine input. */
+data class DebeziumInput(
+    val properties: Map<String, String>,
+    val state: DebeziumState,
+    val isSynthetic: Boolean,
+)
+
+/** Debezium Engine output, other than records of course. */
+data class DebeziumState(
+    val offset: DebeziumOffset,
+    val schemaHistory: DebeziumSchemaHistory?,
+)
+
+/** Stateless connector-specific Debezium operations. */
+interface DebeziumOperations<T : Comparable<T>> :
+    CdcPartitionsCreatorDebeziumOperations<T>, CdcPartitionReaderDebeziumOperations<T>
+
+interface CdcPartitionsCreatorDebeziumOperations<T : Comparable<T>> {
+
+    /** Extracts the WAL position from a [DebeziumOffset]. */
+    fun position(offset: DebeziumOffset): T
+
+    /** Synthesizes a [DebeziumInput] when no incumbent [OpaqueStateValue] is available. */
+    fun synthesize(): DebeziumInput
+
+    /** Builds a [DebeziumInput] using an incumbent [OpaqueStateValue]. */
+    fun deserialize(opaqueStateValue: OpaqueStateValue, streams: List<Stream>): DebeziumInput
+}
+
+interface CdcPartitionReaderDebeziumOperations<T : Comparable<T>> {
+
+    /** Transforms a [DebeziumRecordValue] into an [AirbyteRecordMessage]. */
+    fun toAirbyteRecordMessage(
+        key: DebeziumRecordKey,
+        value: DebeziumRecordValue
+    ): AirbyteRecordMessage
+
+    /** Maps a [DebeziumState] to an [OpaqueStateValue]. */
+    fun serialize(debeziumState: DebeziumState): OpaqueStateValue
```

Comment:
nit: maybe rename this to `serializeToAirbyteStateMessage` ?

Reply:
It's actually not an AirbyteStateMessage that we're serializing to, instead it's the "shared_state" value, but yes, I'll rename this.

Reply:
Eh, I've changed my mind, I think I'll leave it as-is, otherwise I have to rename `deserialize` also, and I don't know what to do.